### PR TITLE
[Datasets] Return ndarray dicts for single-column tabular datasets.

### LINKED ISSUE #30407: [Datasets] Wrong batch type passed to UDF
### What happened + What you expected to happen

I applied a UDF to a dataset containing structured data. I expected my UDF to receive a dict as input, but I got an ndarray instead.

### Versions / Dependencies

Ray: c7115135ea131b29bd6ff3d32e4f90297e5e770e
OS: M1 Mac
Python: 3.10.8

### Reproduction script

```python
>>> import ray
>>> dataset = ray.data.read_images("s3://air-example-data-2/imagenet-sample-images/")
Dataset(num_blocks=4, num_rows=14, schema={image: ArrowVariableShapedTensorType(dtype=uint8, ndim=3)})
>>> def map_fn(batch):
...     assert isinstance(batch, dict)
...     return batch
>>> dataset.map_batches(map_fn, batch_format="numpy")
Traceback (most recent call last):
  File "/Users/balaji/Documents/GitHub/ray/repro.py", line 11, in <module>
    dataset.map_batches(map_fn, batch_format="numpy")
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 619, in map_batches
    return Dataset(plan, self._epoch, self._lazy)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 225, in __init__
    self._plan.execute(allow_clear_input_blocks=False)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/plan.py", line 314, in execute
    blocks, stage_info = stage(
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/plan.py", line 678, in __call__
    blocks = compute._apply(
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/compute.py", line 151, in _apply
    raise e from None
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/compute.py", line 138, in _apply
    results = map_bar.fetch_until_complete(refs)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/progress_bar.py", line 75, in fetch_until_complete
    for ref, result in zip(done, ray.get(done)):
  File "/Users/balaji/Documents/GitHub/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/_private/worker.py", line 2309, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AssertionError): ray::_map_block_nosplit() (pid=49247, ip=127.0.0.1)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/compute.py", line 484, in _map_block_nosplit
    for new_block in block_fn(blocks, *fn_args, **fn_kwargs):
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 597, in transform
    yield from process_next_batch(batch)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 579, in process_next_batch
    batch = batch_fn(batch, *fn_args, **fn_kwargs)
  File "/Users/balaji/Documents/GitHub/ray/repro.py", line 8, in map_fn
    assert isinstance(batch, dict)
AssertionError
```

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Single-column tabular datasets have a human-readable column name (e.g. "images" for data read using our `ray.data.read_images()`) API, so we should preserve that human-readable column name when using the NumPy batch format for mapper and iterator APIs. This PR does that, ensuring that single-column ndarray dict batches are yielded instead of just ndarrays.

## Before

```python
ds = ray.data.from_items([{"col_name": np.ones((2, 2))}, {"col_name": np.ones((2, 2))}])

def udf(batch: np.ndarray) -> Dict[str, np.ndarray]:
    # batch is a single ndarray, and the "col_name" column name is lost
    # unless we hardcode it back.
    return {"col_name": process(batch)}

ds = ds.map_batches(udf, batch_format="numpy")
```

## After

```python
ds = ray.data.from_items([{"col_name": np.ones((2, 2))}, {"col_name": np.ones((2, 2))}])

def udf(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
    # batch is a single-item dict, preserving the "col_name" column name.
    return {k: process(v) for k, v in batch.items()}

ds = ds.map_batches(udf, batch_format="numpy")
```

## Related issue number

Closes https://github.com/ray-project/ray/issues/30407

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_dataset.py
Comment (bveeramani) [inline python/ray/data/tests/test_dataset.py:586]: Maybe I'm being really dumb, but how is `map_batches` not raising an assertion error? `mapper` asserts that 

```
        assert list(col_arrs.keys()) == ["a", "b", "c"]
```

but the input table only contains a "c" column?

File: python/ray/data/tests/test_dataset.py
Comment (bveeramani) [inline python/ray/data/tests/test_dataset.py:636]: Same here.

File: python/ray/data/tests/test_dataset.py
Comment (bveeramani) [inline python/ray/data/tests/test_dataset.py:1930]: Why do we need a try-catch here?

File: python/ray/data/tests/test_dataset.py
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:586]: Not being dumb, that's a good point! I'm not sure how these tests passed locally for me... I didn't finish porting the tests.

Let me dig into this.

File: python/ray/data/tests/test_dataset.py
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:1930]: To restore the old `actor_prefetcher_enabled` config value if the test fails, otherwise all of the below tests will be run with `context.actor_prefetcher_enabled=False`, which is not the default.

**Review by c21 — COMMENTED** [review]
LGTM for change in `arrow_block.py` and `pandas_block.py`, except test issues raised by @bveeramani.

**Review by jiaodong — COMMENTED** [review]
change looks good to me, besides unit test can you also ensure we've updated all image examples in release tests & examples ? IIUC this change means our preprocessor UDF for image examples should now have 

```python
def preprocess(image_batch: Dict[str, np.ndarray])
```

instead of 

```python
def preprocess(image_batch: np.ndarray)
```
such as 
https://sourcegraph.com/github.com/ray-project/ray/-/blob/release/air_tests/air_benchmarks/workloads/gpu_batch_prediction.py

**Comment by amogkam** [comment]
Thanks @clarkzinzow! Can you merge in master to pick up this change: https://github.com/ray-project/ray/pull/30440.

This will make sure the AIR tests and examples get run on these changes and should address @jiaodong's concerns on updating the examples.

File: python/ray/data/_internal/arrow_block.py
Comment (c21) [inline python/ray/data/_internal/arrow_block.py:248]: why we don't need to check `columns == self._table_column_names` here?

File: python/ray/data/_internal/simple_block.py
Comment (c21) [inline python/ray/data/_internal/simple_block.py:101]: `select()` for simple block should be callable, instead of string. wondering why we need to change here?

File: python/ray/data/_internal/arrow_block.py
Comment (clarkzinzow) [reply python/ray/data/_internal/arrow_block.py:248]: Here, `columns` is a single column string, so we're selecting a single column; this is used internally e.g. in our sort and groupby implementation, and we always want to return a single ndarray in this case. This should never be user-facing, though, all user-facing APIs that eventually use `BlockAccessor.to_numpy()` will either give no argument (for the full table) or a list of columns (for a select-like API).

File: python/ray/data/_internal/simple_block.py
Comment (clarkzinzow) [reply python/ray/data/_internal/simple_block.py:101]: This is to simplify the groupby logic, see the changed in `grouped_dataset.py`: https://github.com/ray-project/ray/pull/30448/files#diff-0e694ceda06b25cb570091a11fc44f31ba8d89ad893f0577946c5b364eaed2cf

**Review by bveeramani — APPROVED** [review]
Looks like `test_batch_mapper` is failing with `ValueError: Column must be a single callable when selecting on Simple blocks, but got: [].`

Otherwise, LGTM

**Comment by clarkzinzow** [comment]
@bveeramani Yeah I saw, although that test has been flaky in master for a while due to a task cancellation bug in Ray Core, so not sure if it needs digging into. There's a fix in the works for the latter, so I might wait until that's merged, rebase, and run again.

**Comment by jiaodong** [comment]
there're still test failures on CI from arrow, are they blockers of merging ?

**Comment by clarkzinzow** [comment]
@jianoaix @c21 pinging for data code owner review!

**Review by c21 — APPROVED** [review]
LGTM

**Comment by clarkzinzow** [comment]
@ray-project/ray-docs Need approval from a docs codeowner!

**Review by richardliaw — APPROVED** [review]

**Pull request merged**
